### PR TITLE
Convert codebase to typescript

### LINKED ISSUE #78: [TypeScript] Add TypeScript support
I think we should write this client in TypeScript. 

If we don't, we should at least add the declaration files for the various object the client returns with Definitely Typed:

- https://basarat.gitbooks.io/typescript/docs/types/@types.html
- https://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html

I started writing some declaration files in the Vue.js dashboard. We can take the files from there as a starting point.

### PR DESCRIPTION
**Pull request by ksol**
Converts the codebase to typescript. Each commit handles one part of the rewrite. [alexjoverm/typescript-library-starter](https://github.com/alexjoverm/typescript-library-starter) was used as guide to setup the whole thing.

1. Adds support for typescript for the test runner
2. Rewrite `src` in TS, touching tests as little as possible
3. Restore linting
4. Ensure build is still working
5. Generate documentation

Most, if not all of the types have a name matching the ones previously supplied via ESdoc. File structure is the same too.

The test files have almost not been touched (only a minor null-check has been added), so if everything is green... this mean this _should_ work.

Examples are also still working, which _should_ mean that the build output is functionally the same.

Docs are still generated, but the output is different than before; I find them slightly less pleasant than before but still exploitable.

---

So, to sum up, what's should be checked before merging:
- that the doc output is satisfying
- that the build output is consumable as it was before for non-TS projects without too many changes (having to slightly change the import statements is ok, having to do more is probably not)
- that the source is consumable for TS project with little to no changes for TS consumers.

If/when all this is a go, I'd suggest bumping to `0.1.0` since this is a non-trivial change that may reveal to be breaking.

Fix #78

### CODE REVIEW

**Comment by johnsudaar** [comment]
Just wow! Thanks a lot !!!! I agree with @EtienneM on this if the only blocking part is the doc, we could merge this and iterate on it later.

For the:

```
that the build output is consumable as it was before for non-TS projects without too many changes (having to slightly change the import statements is ok, having to do more is probably not)
```

Since the project has not been released yet. I don't mind breaking the API. So if other thing changes it's fine by me if it's documented and if it doesn't break the DX.

**Comment by ksol** [comment]
Regarding build output, I think I'll try to end with the following:

- people using this repo as source will have to be able to integrate TS projects. No build outputs will be present
- for packages published on npms, es-modules consumers should be able to use the package as it works currently, wether they support TS or not (meaning js output will be outputed next to the TS ones). CJS (node-style) consumers will be able to require 'dist/scalingo.js' as it is done currently and seen in the examples.

**Comment by ksol** [comment]
Alright, as of last commit there should be no change for consumers importing `scalingo` (since it points to the rollup build).

Consumers wishing to consume either TypeScript or the source directly can require `scalingo/src/index` or `scalingo/src`.

I don't know if that's the best way of doing things but at least it's working.

If that's fine for you @johnsudaar @EtienneM, let's get this merged, bump to `0.1.0` and publish (..and fix the build output with patch levels until it's actually working. 😄)
